src/common/sai.rs

```diff
@@ -0,0 +1,551 @@
+//!
+//! Each data line comes with its own 32x32 FIFO allowing for a full frame to be sent and/or received
+//! without software interaction.
+//!
+//! The configuration of the SAI is encoded in configuration structure that can be used with a singular
+//! configure method.
+
+use crate::iomuxc::{consts, sai};
+use crate::ral;
+use core::marker::PhantomData;
+
+/// Audio word byte order
+#[derive(Clone, Copy, Default, Eq, PartialEq)]
+pub enum ByteOrder {
+    /// Least significant byte first
+    #[default]
+    LSB = 0,
+    /// Most significant byte first
+    MSB = 1,
+}
+
+/// Mode of operation for the SAI peripheral
+#[derive(Clone, Copy, Default, Eq, PartialEq)]
+pub enum Mode {
+    /// Master mode where all clocks are generated from the SAI
+    #[default]
+    Master = 0,
+    /// Slave mode where all audio clocks are expecting to come from external sources
+    Slave = 1,
+    /// Bitclock Master, FrameSync Slave
+    BclkMasterFrameSyncSlave = 2,
+    /// Bitclock Slave, FrameSync Master
+    BclkSlaveFrameSyncMaster = 3,
+}
+
+/// Clock Polarity Options for Bclk/Mclk
+#[derive(Clone, Copy, Default, Eq, PartialEq)]
+pub enum ClockPolarity {
+    /// Transmitted clock implies active high
+    #[default]
+    ActiveHigh = 0,
+    /// Transmitted clock implies active low
+    ActiveLow = 1,
+}
+
+#[allow(non_upper_case_globals)]
+impl ClockPolarity {
+    /// Received clock implies sample on rising edge
+    pub const SampleOnRising: ClockPolarity = ClockPolarity::ActiveLow;
+    /// Received clock implies sample on falling edge
+    pub const SampleOnFalling: ClockPolarity = ClockPolarity::ActiveHigh;
+}
+
+/// Mclk source option
+#[derive(Clone, Copy, Default, Eq, PartialEq)]
+pub enum MclkSource {
+    /// Mclk sourced from system clock
+    #[default]
+    Sysclk = 0,
+    /// Select 1, part dependent
+    Select1 = 1,
+    /// Select 2, part dependent
+    Select2 = 2,
+    /// Select 3, part dependent
+    Select3 = 3,
+}
+
+/// Frame sync mode between rx/tx
+#[derive(Clone, Copy, Default, Eq, PartialEq)]
+pub enum SyncMode {
+    /// Both tx/rx are setup as being independent of eachother for frame sync
+    #[default]
+    Async = 0,
+    /// Tx synchronously follows Rx frame sync
+    TxFollowRx = 1,
+    /// Rx synchronously follows Tx frame sync
+    RxFollowTx = 2,
+}
+
+/// Frame Sync Width
+#[derive(Clone, Copy, Default, Eq, PartialEq)]
+pub enum SyncWidth {
+    /// Frame sync width is the size of the word
+    #[default]
+    WordSize,
+}
+
+/// Source for Bclk, check part datasheet for details
+#[derive(Clone, Copy, Default, Eq, PartialEq)]
+pub enum BclkSource {
+    /// Bus clock is the source of Bclk
+    #[default]
+    Bus = 0,
+    /// Option 1, part dependent
+    Opt1 = 1,
+    /// Option 2, part dependent
+    Opt2 = 2,
+    /// Option 3, part dependent
+    Opt3 = 3,
+}
+
+/// Packing is useful to have function variants over so is provided as a trait
+pub trait Packing<const WORD_SIZE: u8> {
```

Comment:
Should this trait be sealed to stop users from providing arbitrary `fpack` values?

Reply:
Done!